src/main/java/org/truffleruby/language/LazyRubyNode.java

```diff
@@ -63,11 +63,13 @@ public RubyNode resolve() {
                 Log.LOGGER.info(() -> "lazy translating " + RubyLanguage.fileLine(getParent().getEncapsulatingSourceSection()) + " in " + getRootNode());
             }
 
-            resolved = resolver.get();
-            transferFlagsTo(resolved);
+            final RubyNode result = resolver.get();
+            transferFlagsTo(result);
 
-            replace(resolved, "lazy node resolved");
-            return resolved;
+            // publish
```

Comment:
> I could use an explicit barrier if you think that's better.

Yeah I think that would be much easier to read.

> The proper, clean solution would be to always keep the LazyRubyNode around

Do you mean by having the new node as a child?